src/c_api.cpp

```diff
@@ -422,6 +433,214 @@ class Booster {
+                     bool* is_data_float32_ptr, int num_matrices) {
+    auto predictor = CreatePredictor(num_iteration, predict_type, ncol, config);
+    auto pred_sparse_fun = predictor.GetPredictSparseFunction();
+    std::vector<std::vector<std::unordered_map<int, double>>>& agg = *agg_ptr;
+    OMP_INIT_EX();
+    #pragma omp parallel for schedule(static)
+    for (int i = 0; i < nrow; ++i) {
+      OMP_LOOP_EX_BEGIN();
+      auto one_row = get_row_fun(i);
+      agg[i] = std::vector<std::unordered_map<int, double>>(num_matrices);
+      pred_sparse_fun(one_row, &agg[i]);
+      OMP_LOOP_EX_END();
+    }
+    OMP_THROW_EX();
+    // calculate the nonzero data and indices size
+    int elements_size = 0;
+    for (int i = 0; i < static_cast<int>(agg.size()); i++) {
+      auto row_vector = agg[i];
+      for (int j = 0; j < static_cast<int>(row_vector.size()); j++) {
+        elements_size += row_vector[j].size();
+      }
+    }
+    *out_elements_size = elements_size;
+    *is_data_float32_ptr = false;
+    // allocate data and indices arrays
+    if (data_type == C_API_DTYPE_FLOAT32) {
+      *out_data = new float[elements_size];
+      *is_data_float32_ptr = true;
+    } else if (data_type == C_API_DTYPE_FLOAT64) {
+      *out_data = new double[elements_size];
+    } else {
+      Log::Fatal("Unknown data type in PredictSparse");
+      return;
+    }
+    *out_indices = new int32_t[elements_size];
+  }
+
+  void PredictSparseCSR(int num_iteration, int predict_type, int nrow, int ncol,
+                        std::function<std::vector<std::pair<int, double>>(int row_idx)> get_row_fun,
+                        const Config& config,
+                        int64_t* out_len, void** out_indptr, int indptr_type,
+                        int32_t** out_indices, void** out_data, int data_type) {
+    std::lock_guard<std::mutex> lock(mutex_);
+    // Get the number of trees per iteration (for multiclass scenario we output multiple sparse matrices)
+    int num_matrices = boosting_->NumModelPerIteration();
+    bool is_indptr_int32 = false;
+    bool is_data_float32 = false;
+    int indptr_size = (nrow + 1) * num_matrices;
+    if (indptr_type == C_API_DTYPE_INT32) {
+      *out_indptr = new int32_t[indptr_size];
+      is_indptr_int32 = true;
+    } else if (indptr_type == C_API_DTYPE_INT64) {
+      *out_indptr = new int64_t[indptr_size];
+    } else {
+      Log::Fatal("Unknown indptr type in PredictSparseCSR");
+      return;
+    }
+    // aggregated per row feature contribution results
+    std::vector<std::vector<std::unordered_map<int, double>>> agg(nrow);
+    int elements_size = 0;
+    PredictSparse(num_iteration, predict_type, nrow, ncol, get_row_fun, config, &elements_size, &agg,
+                  out_indices, out_data, data_type, &is_data_float32, num_matrices);
+    // copy vector results to output for each row
+    // TODO(imatiach-msft): should this be parallelized, or is there no perf benefit?
```

Comment:
you can parallel the copy of `out_data` and `out_indptr'.

Reply:
will do, working on this now, running into some indexing errors in the parallel code

Reply:
done, parallelized the CSR matrix code